spacy/tokens/doc.pyx

```diff
@@ -1734,6 +1735,89 @@ cdef class Doc:
+        affix_lengths: an integer list specifying the lengths of affixes to be hashed. For example, if *affix_lengths==[2, 3]*,
+            *suffs_not_prefs==True* and *case_sensitive==True*, the suffixes hashed for "spaCy" would be "Cy" and "aCy".
+        search_chars: a string containing characters to search for within each token, starting at the beginning or end depending on the
+            value of *suffs_not_prefs*.
+        search_lengths: an integer list specifying the lengths of search results to be hashed. For example if *search_lengths==[1, 2]*,
+            *search_chars=="aC", *suffs_not_prefs==True* and *case_sensitive==True*, the searched strings hashed for "spaCy" would be
+            "C" and "Ca".
+        
+        For a document with tokens ["spaCy", "and", "Prodigy"], the NumPy array returned by 
+        *get_affix_hashes(True, True, [2, 4, 6], "yC", [1, 2])* would correspond to
+
+        [[hash("Cy"), hash("paCy"), hash("spaCy"), hash("y"), hash("yC")],
+        [hash("nd"), hash("and", hash("and"), hash(" "), hash("  "))],
+        [hash("gy"), hash("digy"), hash("rodigy"), hash("y"), hash("y ")]]
+
+        UTF-16 is used to encode the token texts, as this results in two-byte representations for all characters that are realistically
+        interesting when learning features from words. UTF-16 can also contain four-byte representations, but neither of the byte pairs in 
+        a four-byte representation is ever valid in its own right as a two-byte representation. In the rare case that a four-byte 
+        representation occurs in a string being analysed, each of its two-byte pairs is treated as a separate character. A four-byte
+        representation in *search_chars*, on the other hand, is not supported and results in a ValueError(E1046).
+        """
+
+        cdef const unsigned char[:] search_chars_v = _get_utf16_memoryview(search_chars, True)
+        cdef unsigned int longest_search_length = max(search_lengths) if len(search_lengths) > 0 else 0         
+        cdef bytes found_char_buf_bytes = (bytes(" " * longest_search_length, "UTF-16"))[2:] # first two bytes express endianness
+        cdef char* found_char_buf = found_char_buf_bytes
+        cdef unsigned int search_chars_v_len = len(search_chars_v), found_char_buf_len = len(found_char_buf_bytes)
+        
+        cdef unsigned int num_toks = len(self), num_norm_hashes = len(affix_lengths), num_search_hashes = len(search_lengths)
+        cdef np.ndarray[np.int64_t, ndim=2] hashes = numpy.empty((num_toks, num_norm_hashes + num_search_hashes), dtype="int64")
+
+        cdef const unsigned char[:] tok_str_v
+        cdef unsigned int tok_idx, tok_str_v_len, hash_idx, affix_start, char_comb_len
+        cdef attr_t num_tok_attr
+        cdef str str_tok_attr
+        
+        for tok_idx in range(num_toks):
+            num_tok_attr = self.c[tok_idx].lex.orth if case_sensitive else self.c[tok_idx].lex.lower
```

Comment:
I tried various ways of retrieving the string for each token. Perhaps somewhat counterintuitively this is the fastest I came up with, although there may well be a better one I haven't tried. It's certainly faster than retrieving the character index and length of each token and slicing `doc.text`. Declaring a memory view on `doc.text` is unfortunately not an option because there is no guarantee that all characters will be the same width.